feat: add nsq scaler

Ulminator opened the pull request:
This change adds a new scaler for [NSQ](https://github.com/nsqio/nsq), enabling auto-scaling based on topic/channel depth. The main components of NSQ that the scaler interacts with are [nsqlookupd](https://nsq.io/components/nsqlookupd.html) and [nsqd](https://nsq.io/components/nsqd.html). First the scaler makes [/lookup](https://nsq.io/components/nsqlookupd.html#get-lookup) requests to the `nsqlookupd` instances in the cluster to determine which `nsqd` instances have messages for the given topic. Then for each `nsqd` instance the scaler makes [/stats](https://nsq.io/components/nsqd.html#get-stats) requests to and aggregates the results to determine a final depth value.

NSQ consumers create the channel they are reading if they do not exist. As such, if the channel does not exist on an `nsqd` host, the topic depth is used. The consumer pods that KEDA would be responsible in spawning would then create the channel.

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #3281 

Relates to:
- kedacore/test-tools#175
- kedacore/test-tools#184
- kedacore/keda-docs#1485

Ulminator commented:
@JorTurFer thanks for merging in kedacore/test-tools#175! I updated the code here to utilize the image that can be created from that PR.

I also wanted to let you know that I have created the corresponding docs PR as well here: kedacore/keda-docs#1485

The last remaining TODO's seem to just be getting these CI checks to pass. Is it safe to ignore the remaining `semgrep` checks since they just relate to tests? I initially got flagged for `no-direct-write-to-responsewriter` and now I'm getting `no-fprintf-to-responsewriter`, yet I do see both of these patterns being used in other test files.

zroubalik commented on CHANGELOG.md:
```diff
@@ -55,6 +55,8 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 
 ## Unreleased
 
+- **General:** Introduce new NSQ Scaler ([#3281](https://github.com/kedacore/keda/issues/3281))
```
Move this please below to `New` section

zroubalik commented:
/run-e2e nsq

Ulminator commented:
I see the e2e tests have failed, but I'm wondering if the e2e test image was created? 🤔 I don't see it [here](https://github.com/orgs/kedacore/packages?type=source). I'm guessing if `make e2e-images` is run in the `test-tools` repo it should create the `ghcr.io/kedacore/tests-nsq:latest` image. I'm thinking the issue was an image pull error due to it not existing yet.

```
# Local run w output snippet from describing a consumer pod
  Warning  Failed     9s    kubelet            Failed to pull image "ghcr.io/kedacore/tests-nsq:latest": failed to pull and unpack image "ghcr.io/kedacore/tests-nsq:latest": failed to resolve reference "ghcr.io/kedacore/tests-nsq:latest": failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://ghcr.io/token?scope=repository%3Akedacore%2Ftests-nsq%3Apull&service=ghcr.io: 401 Unauthorized
  Warning  Failed     9s    kubelet            Error: ErrImagePull
  Normal   BackOff    8s    kubelet            Back-off pulling image "ghcr.io/kedacore/tests-nsq:latest"
  Warning  Failed     8s    kubelet            Error: ImagePullBackOff
```

I also noticed I had only updated one of the refs of my custom images to `ghcr.io/kedacore/tests-nsq:latest` so I went ahead and made that fix.

JorTurFer commented:
The image was created but GH creates them as private by default, now it's public

JorTurFer commented:
/run-e2e nsq

JorTurFer commented:
Do we need to update the policy to include any extra permission? https://github.com/kedacore/testing-infrastructure/blob/main/terraform/modules/aws/iam/main.tf#L88-L96

Ulminator commented:
> Do we need to update the policy to include any extra permission? https://github.com/kedacore/testing-infrastructure/blob/main/terraform/modules/aws/iam/main.tf#L88-L96
> 
> If yes, could you open a PR there?

No additional permissions are required 👍

Ulminator commented:
@JorTurFer just wanted to check in and see if you need anything else before merging this in?

JorTurFer commented on pkg/scalers/nsq_scaler.go:
```diff
@@ -0,0 +1,358 @@
+		return fmt.Errorf("no nsqLookupdHTTPAddresses given")
+	}
+
+	if m.Topic == "" {
+		return fmt.Errorf("no topic given")
+	}
+
+	if m.Channel == "" {
+		return fmt.Errorf("no channel given")
+	}
```
Are these checks required @wozniakjan ? The entry isn't optional

Ulminator replied:
Those seemed fine to remove 👍 I tested removing `channel` from the ScaledObject in the nsq e2e test and saw logs like this in the controller:

```
2024-11-05T00:06:15Z    ERROR   scale_handler   error resolving auth params     {"type": "ScaledObject", "namespace": "nsq-test-ns", "name": "nsq-test-so", "triggerIndex": 0, "error": "error parsing NSQ metadata: error parsing nsq metadata: missing required parameter \"channel\" in [triggerMetadata resolvedEnv]"}
```

JorTurFer commented on pkg/scalers/nsq_scaler.go:
```diff
@@ -0,0 +1,358 @@
+		return []external_metrics.ExternalMetricValue{}, false, err
+	}
+
+	s.logger.Info("GetMetricsAndActivity", "metricName", metricName, "depth", depth)
```
This should be a debug info
```suggestion
	s.logger.V(1).Info("GetMetricsAndActivity", "metricName", metricName, "depth", depth)
```

JorTurFer commented on pkg/scalers/nsq_scaler.go:
```diff
@@ -0,0 +1,358 @@
+	}
+
+	if len(nsqdHosts) == 0 {
+		s.logger.Info("no nsqd hosts found for topic", "topic", s.metadata.Topic)
```
This should be a debug info:
```suggestion
		s.logger.V(1).Info("no nsqd hosts found for topic", "topic", s.metadata.Topic)
```

JorTurFer commented on pkg/scalers/nsq_scaler.go:
```diff
@@ -0,0 +1,358 @@
+				if ch.Paused {
+					// if it's paused on a single nsqd host, it's depth should not go into the aggregate
+					// meaning if paused on all nsqd hosts => depth == 0
+					s.logger.Info("channel is paused", "topic", topic, "channel", channel, "host", result.host)
```
This should be a debug info:
```suggestion
					s.logger.V(1).Info("channel is paused", "topic", topic, "channel", channel, "host", result.host)
```

JorTurFer commented on pkg/scalers/nsq_scaler.go:
```diff
@@ -0,0 +1,358 @@
+			}
+			if !channelExists {
+				// topic exists with channels, but not the one in question - fallback to topic depth
+				s.logger.Info("channel does not exist for topic", "topic", topic, "channel", channel, "host", result.host)
```
This should be a debug info:
```suggestion
				s.logger.V(1).Info("channel does not exist for topic", "topic", topic, "channel", channel, "host", result.host)
```

JorTurFer commented on pkg/scalers/nsq_scaler.go:
```diff
@@ -0,0 +1,358 @@
+	return &nsqScaler{
+		metricType: metricType,
+		metadata:   nsqMetadata,
+		httpClient: kedautil.CreateHTTPClient(config.GlobalHTTPTimeout, true),
```
The `true` here is to ignore the TLS certificate verification. We should expose the field to users and set it `false` as default

Ulminator replied:
I added the `unsafeSsl` field to help enable this. I was also just using `http` previously, so it seemed like I needed to add something like `useHttps` as well.

JorTurFer commented on tests/scalers/nsq/nsq_test.go:
```diff
@@ -0,0 +1,214 @@
+spec:
+  pollingInterval: 5
+  cooldownPeriod: 10
+  idleReplicaCount: 0
```
This field isn't required
```suggestion
```

JorTurFer commented on tests/scalers/nsq/nsq_test.go:
```diff
@@ -0,0 +1,214 @@
+	AssertReplicaCountNotChangeDuringTimePeriod(t, kc, deploymentName, testNamespace, 0, 60)
+}
+
+func testScaleOut(t *testing.T, kc *kubernetes.Clientset, data templateData) {
+	t.Log("--- testing scale out ---")
+
+	data.MessageCount = 1 // 5 already published + 1 > activationDepthThreshold
+	KubectlReplaceWithTemplate(t, data, "jobTemplate", jobTemplate)
+
+	require.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 1, 60, 1),
+		"replica count should be 1 after 1 minute")
+}
```
We should increase the queue to scale from 1 to at least 2 to verify that the scaling phase also works. My suggestion at this point is setting `maxReplicaCount:2` and updating the consumer to introduce a sleep of 1 second or so after getting each message. This allows to enqueue 30-60 extra messages, ensuring that the workload scales to 2 instances

Ulminator replied:
I was able to modify the test image here [kedacore/test-tools/#184](https://github.com/kedacore/test-tools/pull/184). The e2e tests should be better now 👍

JorTurFer reviewed (commented):
Looking good. I've kept some comments inline.
I'm worried about the lifecycle of the project, I closed the issue with the feature request a year and a half ago because there wasn't activity, and it looked as no longer maintained. 
@zroubalik @wozniakjan

Ulminator commented:
@JorTurFer thanks for taking a look! I believe I addressed all of your comments, but let me know if anything else needs to change of course.

> I'm worried about the lifecycle of the project, I closed the issue with the feature request a year and a half ago because there wasn't activity, and it looked as no longer maintained.

I'm an engineer at [Bitly](https://bitly.com/), which is where NSQ was initially developed. We use it extensively throughout our stack, and while updates to the project are not as frequent as they were in the past, it is stable. There are plans for updates in the near future and it is under active development.

JorTurFer commented on tests/scalers/nsq/nsq_test.go:
```diff
@@ -0,0 +1,225 @@
+	maxReplicaCount          = 2
+	topicName                = "test_topic"
+	channelName              = "test_channel"
+	depthThreshold           = 10
```
```suggestion
	depthThreshold           = 1
```

JorTurFer commented on tests/scalers/nsq/nsq_test.go:
```diff
@@ -0,0 +1,225 @@
+	t.Log("--- testing scale out ---")
+
+	// can handle depthThreshold messages per replica - using maxReplicaCount + 1 to ensure scaling to maxReplicaCount
+	data.MessageCount = depthThreshold * (maxReplicaCount + 1)
```
We don't need to be fine with this, to ensure that the scaling happens, we can push a bigger amount.
The consumer pulls 1 message per second, so we can publish 80 messages, which means 40 seconds approx to be consumed (which is totally okey). This also gives some time in case of pulling image is required
```suggestion
	data.MessageCount = 80
```

JorTurFer commented:
/run-e2e nsq

JorTurFer commented on tests/scalers/nsq/nsq_test.go:
```diff
@@ -0,0 +1,224 @@
+	KubectlReplaceWithTemplate(t, data, "jobTemplate", jobTemplate)
+	AssertReplicaCountNotChangeDuringTimePeriod(t, kc, deploymentName, testNamespace, 0, 20)
+
+	data.MessageCount = 1 // total message count > activationDepthThreshold
```
```suggestion
	data.MessageCount = 1
```

JorTurFer reviewed (approved):
LGTM! Awesome job! 🙇

JorTurFer commented:
@zroubalik @wozniakjan PTAL

JorTurFer commented:
/run-e2e nsq

wozniakjan reviewed (approved):
lgtm, thank you!

not sure what is wrong with `Semgrep / Analyze Semgrep` action, it doesn't allow me to see anything in there

Ulminator commented:
@JorTurFer, I have updated this PR along with [kedacore/docs#1485](https://github.com/kedacore/keda-docs/pull/1485) to account for the 2.16 release that happened last week, PTAL!

JorTurFer commented:
/run-e2e nsq

JorTurFer commented:
Thanks for your contribution!
